### PR TITLE
NI Traktor Kontrol S2 MK1 mapping

### PR DESCRIPTION
**Pull request by leifhelm**
I added a mapping for the Native Instrument Traktor Kontrol S2 DJ controller in [#3905](https://github.com/mixxxdj/mixxx/pull/3905).
This is the documentation of that mapping.
It is based on that of the Traktor Kontrol S2 MK2.

### CODE REVIEW

File: source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+to the outputs. The Kontrol S2 MK1 can run with only USB
+bus power and an optional power supply can be connected to make the LEDs
+brighter and the headphone output louder.
+
+The Kontrol S2 MK1 can be distinguished from the Mk2 by the jog wheels.
+The top of the jog wheels on the Mk1 are black; The top of the jog wheels
+on the MK2 are shiny aluminum; The Kontrol S2 Mk3 does not have
+effects knobs at the top.
+
+.. versionadded:: 2.3.1
```
Comment (JoergAtGithub) [inline source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst:18]: 2.3.4

File: source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    Adjusts the :term:`tempo`.
+:hwlabel:`RESET` button
+    Mixxx `does not yet have a very useful flux/slip
+    mode <https://bugs.launchpad.net/mixxx/+bug/1475303>`__, so instead this
+    button toggles the top pad row to the intro/outro cues. Press the button
+    when it is lit to return the top pad row to hotcue mode.
+
+    When slip mode is implemented in Mixxx in the future, this could be
+    changed so :hwlabel:`SHIFT` + :hwlabel:`RESET` button activates intro & outro mode for the
+    pads.
```
Comment (JoergAtGithub) [inline source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst:124]: Mixxx has a Slip mode, which of cause behaves different than any other DJ software implements it.
Please describe just what happens in Mixxx 2.3 e.g. :https://deploy-preview-412--mixxx-manual.netlify.app/chapters/appendix/mixxx_controls.html?highlight=slip#control-[ChannelN]-slip_enabled

Discussions about future features/changes shouldn't be in the manual.

File: source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+:hwlabel:`LEVEL` meters
+    The meters show the levels for each deck.
+Crossfader
+    Crossfade between the decks.
+
+Effects
+~~~~~~~
+
+The Kontrol S2 MK1 uses the `standard Mixxx effects
+mapping <https://github.com/mixxxdj/mixxx/wiki/standard-effects-mapping>`__.
```
Comment (JoergAtGithub) [inline source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst:175]: The manual should be self contained and should not link into the Wiki. Please copy what applies for this mapping.

**Review by JoergAtGithub — CHANGES_REQUESTED** [review]

File: source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst (hunk 2 above)
Comment (leifhelm) [reply source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst]: This is a direct copy from the Mk2 manual.

File: source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst (hunk 3 above)
Comment (leifhelm) [reply source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst]: It is the same in the Mk2 manual

Comment (leifhelm) [reply source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst]: I made it self contained instead is that fine to?

**Comment by leifhelm** [comment]
Where should I add Info about the adjustment of the `JogWheelTouchThreshold`?

**Comment by JoergAtGithub** [comment]
> Where should I add Info about the adjustment of the `JogWheelTouchThreshold`?

Are the values of the Native Instruments Controller Editor stored in the device? https://support.native-instruments.com/hc/de/article_attachments/205618865/Jog_Wheel_Calibration_Guide_English.pdf

Access to device internal calibration data would be the typical use case of an HID FeatureReport. Does the S2 list FeatureReports in the HID ReportDescriptor?

BTW: NI calibrates different thresholds for 4 positions of each JogWheel. Your mapping considers only one threshold per JogWheel.

File: source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+1/4" TRS outputs and a pair of unbalanced RCA outputs which both output
+the main mix, a 1/4" TRS headphone jack, and a 1/4" TRS microphone
+input. The microphone input is not digitized and routed internally
+to the outputs. The Kontrol S2 MK1 can run with only USB
+bus power and an optional power supply can be connected to make the LEDs
+brighter and the headphone output louder.
+
+The Kontrol S2 MK1 can be distinguished from the Mk2 by the jog wheels.
+The top of the jog wheels on the Mk1 are black; The top of the jog wheels
+on the MK2 are shiny aluminum; The Kontrol S2 Mk3 does not have
```
Comment (JoergAtGithub) [inline source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst:15]:
```suggestion
The S2 MK1 has pressure sensitive jog wheels made of black plastic, while the S2 MK2 has touch sensitive jog wheels made of shiny aluminium; The Kontrol S2 MK3 does not have
```

File: source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+Decks
+~~~~~
+
+Jog Wheels
+^^^^^^^^^^
+
+Touch the top of the jog wheel and turn it to scratch. Move the jog
+wheel from the edge without touching the top to nudge the track. Hold
+:hwlabel:`SHIFT` and spin the jog wheel to seek quickly.
```
Comment (JoergAtGithub) [inline source/hardware/controllers/native_instruments_traktor_kontrol_s2_mk1.rst:52]:
```suggestion
Press the top of the jog wheel and turn it to scratch. Move the jog
wheel from the edge without pressing the top to nudge the track. Hold
:hwlabel:`SHIFT` and spin the jog wheel to seek quickly.
```

**Comment by leifhelm** [comment]
<details>
<summary>This is the report description</summary>

```
Usage Page (FF01h),                     ; FF01h, vendor-defined
Usage (00h),
Collection (Application),
    Usage (01h),
    Collection (Logical),
        Report ID (1),
        Usage (31h),
        Usage (31h),
        Logical Minimum (-2147483648),
        Logical Maximum (2147483647),
        Report Size (32),
        Report Count (2),
        Input (Variable),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Usage (02h),
        Logical Minimum (0),
        Logical Maximum (1),
        Report Size (1),
        Report Count (56),
        Input (Variable),
    End Collection,
    Usage (02h),
    Collection (Logical),
        Report ID (2),
        Usage (03h),
        Usage (03h),
        Usage (03h),
        Usage (03h),
        Usage (03h),
        Usage (03h),
        Usage (03h),
        Usage (03h),
        Logical Minimum (0),
        Logical Maximum (15),
        Report Size (4),
        Report Count (8),
        Input (Variable),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Logical Minimum (0),
        Logical Maximum (4095),
        Report Size (16),
        Report Count (24),
        Input (Variable),
    End Collection,
    Usage (80h),
    Collection (Logical),
        Report ID (128),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Usage (81h),
        Logical Minimum (0),
        Logical Maximum (31),
        Report Count (61),
        Report Size (8),
        Output (Variable),
    End Collection,
    Usage (01h),
    Collection (Logical),
        Report ID (240),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Usage (04h),
        Logical Minimum (0),
        Logical Maximum (255),
        Report Size (8),
        Report Count (24),
        Feature (Variable),
    End Collection,
    Usage (D0h),
    Collection (Logical),
        Report ID (208),
        Usage (D1h),
        Logical Minimum (0),
        Logical Maximum (255),
        Report Size (8),
        Report Count (32),
        Feature (Variable),
    End Collection,
    Usage (D0h),
    Collection (Logical),
        Report ID (209),
        Usage (D1h),
        Logical Minimum (0),
        Logical Maximum (255),
        Report Size (8),
        Report Count (32),
        Feature (Variable),
    End Collection,
    Usage (D0h),
    Collection (Logical),
        Report ID (210),
        Usage (D1h),
        Logical Minimum (0),
        Logical Maximum (255),
        Report Size (8),
        Report Count (32),
        Feature (Variable),
    End Collection,
    Usage (D0h),
    Collection (Logical),
        Report ID (211),
        Usage (D1h),
        Logical Minimum (0),
        Logical Maximum (255),
        Report Size (8),
        Report Count (32),
        Feature (Variable),
    End Collection,
    Usage (D0h),
    Collection (Logical),
        Report ID (212),
        Usage (D1h),
        Logical Minimum (0),
        Logical Maximum (255),
        Report Size (8),
        Report Count (32),
        Feature (Variable),
    End Collection,
    Usage (D0h),
    Collection (Logical),
        Report ID (213),
        Usage (D1h),
        Logical Minimum (0),
        Logical Maximum (255),
        Report Size (8),
        Report Count (32),
        Feature (Variable),
    End Collection,
    Usage (D0h),
    Collection (Logical),
        Report ID (214),
        Usage (D1h),
        Logical Minimum (0),
        Logical Maximum (255),
        Report Size (8),
        Report Count (32),
        Feature (Variable),
    End Collection,
    Usage (D0h),
    Collection (Logical),
        Report ID (215),
        Usage (D1h),
        Logical Minimum (0),
        Logical Maximum (255),
        Report Size (8),
        Report Count (32),
        Feature (Variable),
    End Collection,
End Collection
```
</details>

**Comment by JoergAtGithub** [comment]
> I do not know what to do with it.

Your device has multiple FeatureReports, which might store calibration data of Faders, Knobs and the Jog:
FeatureReport 0xF0 (240) with 24 Bytes
FeatureReport 0xD0-0xD7 (208-215) with 32 Bytes each

You wrote in the other PR, that your knobs are not centered - maybe the calibration procedure of Traktor Pro might fix this.
Mixxx 2.3 doesn't support FeatureReports, but 2.4 developmend builds do.

**Comment by leifhelm** [comment]
# Feature Report Description

This is what i could figure out about the feature reports:

Feature Report `208` (`0xd0`)
Byte|Endianness|Description
---|---|---
0|-|Always `d0`
1..12|-|I don't know/always `01 00 00 00 10 00 f0 0f 10 00 f0 0f`
13..14|LE|Channel 1 volume fader bottom
15..16|LE|Channel 1 volume fader top
17..18|LE|Channel 2 volume fader bottom
19..20|LE|Channel 2 volume fader top
21..22|LE|Crossfader left
23..24|LE|Crossfader right
25..32|-|Padding (`ff`)

Feature Report `209` (`0xd1`)
Byte|Endianness|Description
---|---|---
0|-|Always `d1`
1..2|LE|Channel 1 FX dry/wet knob left
3..4|LE|Channel 1 FX dry/wet knob center
5..6|LE|Channel 1 FX dry/wet knob right
7..8|LE|Channel 1 FX 1 knob left
9..10|LE|Channel 1 FX 1 knob center
11.12|LE|Channel 1 FX 1 knob right
13..14|LE|Channel 1 FX 2 knob left
15..16|LE|Channel 1 FX 2 knob center
17..18|LE|Channel 1 FX 2 knob right
19..20|LE|Channel 1 FX 3 knob left
21..22|LE|Channel 1 FX 3 knob center
23..24|LE|Channel 2 FX 3 knob right
25..26|LE|Channel 1 EQ hi knob left
27..28|LE|Channel 1 EQ hi knob center
29..30|LE|Channel 1 EQ hi knob right
31..32|LE|Channel 1 EQ mid knob left

Feature Report `210` (`0xd2`)
Byte|Endianness|Description
---|---|---
0|-|Always `d2`
1..2|LE|Channel 1 EQ mid knob center
3..4|LE|Channel 1 EQ mid knob right
5..6|LE|Channel 1 EQ low knob left
7..8|LE|Channel 1 EQ low knob center
9..10|LE|Channel 1 EQ low knob right
11..12|LE|Channel 2 EQ hi knob left
13..14|LE|Channel 2 EQ hi knob center
15..16|LE|Channel 2 EQ hi knob right
17..18|LE|Channel 2 EQ mid knob left
19..20|LE|Channel 2 EQ mid knob center
21..22|LE|Channel 2 EQ mid knob right
23..24|LE|Channel 2 EQ low knob left
25..26|LE|Channel 2 EQ low knob center
27..28|LE|Channel 2 EQ low knob right
29..30|LE|Sample knob left
31..32|LE|Sample knob center

Feature Report `211` (`0xd3`)
Byte|Endianness|Description
---|---|---
0|-|Always `d3`
1..2|LE|Sample knob right
3..4|LE|Channel 2 FX dry/wet knob left
5..6|LE|Channel 2 FX dry/wet knob center
7..8|LE|Channel 2 FX dry/wet knob right
9..10|LE|Channel 2 FX 1 knob left
11..12|LE|Channel 2 FX 1 knob center
13..14|LE|Channel 2 FX 1 knob right
15..16|LE|Channel 2 FX 2 knob left
17..18|LE|Channel 2 FX 2 knob center
19..20|LE|Channel 2 FX 2 knob right
21..22|LE|Channel 2 FX 3 knob left
23..24|LE|Channel 2 FX 3 knob center
25..26|LE|Channel 2 FX 3 knob right
27..32|-|Padding (`ff`)

Feature Report `212` (`0xd4`)
Byte|Endianness|Description
---|---|---
0|-|Always `d4`
1..2|BE|Left jogwheel unpressed
3..4|BE|Left jogwheel pressed
5..6|BE|Right jogwheel unpressed
7..8|BE|Right jogwheel pressed
9|-|Left jogwhell calibration type (`00` user, `ff` factory)
10|-|Right jogwhell calibration type (`00` user, `ff` factory)
11..32|-|Padding (`ff`)

Feature Report `215` (`0xd7`) is factory calibration data for jogwheels:

Restore of both jogwheels to factory settings:
Read 215:  `d7 00 00 00 00 4e c2 26 82 0c 1b 0d b2 0c 9a 0e 16 2a ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff`
Write 212: `d4 0c 1b 0c e6 0c 9a 0d 58 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff`
My guess:
Byte|Endianness|Description
---|---|---
0|-|Always `d7`
1..8|-|I don't know
9..10|BE|Left jogwheel unpressed
11..12|BE|Left jogwheel fully pressed
13..14|BE|Right jogwheel unpressed
15..16|BE|Right jogwheel fully pressed
17|-|`42` i guess
18..32|-|Padding (`ff`)

`212`'s unpressed seems to be the same as unpressed. `212`'s pressed is `(pressed + unpressed) / 2`.

**Comment by JoergAtGithub** [comment]
Great, this table is really valueable information! I think the best place for this technical details is a comment section in the mapping file itself.
What does not fit to my expectations is, the number of calibration data for jog press. The official calibration procedure of Native Instruments is calibration at 4 positions of each jog. This means, there should be 8 pairs of pressed/unpressed values in the calibration data.
Can you confirm, that the calibration data are store in non-volatile memory in the controller?

**Comment by leifhelm** [comment]
My guess is that there is only one force sensor in the jogwheel. The 4 presses do not have to be on the described positions. Pressing in the center works as well. I think the procedure mandates 4 presses to have more data points for the threshold of press/unpressed.

**Comment by JoergAtGithub** [comment]
I guess it is the center component on the printed circuit board here: https://youtu.be/YO15WcFmjko?t=8

**Comment by leifhelm** [comment]
I have taken mine apart and the sensor seems to be an IR distance sensor. So it makes sense that only one value is measured.

**Comment by JoergAtGithub** [comment]
And does your controller keep the calibration data if you disconnect it from power? Than the S2 MK1 would have a non-volatile memory, and you could document in the manual, that users can use the Native Instruments Controller Editor for calibration.
In future, it would be nice to add a calibration feature to Mixxx, as in Traktor Pro.

**Comment by leifhelm** [comment]
The memory of the calibration is non-volatile. The current merge request targets 2.3 which has no capability of reading Feature Reports. I would make a new PR that targets 2.4 with all the stuff only possible in that version, if this is merged like you suggested. Is there anything that is blocking the PRs from beeing merged?

**Comment by JoergAtGithub** [comment]
Could you please:
- Add a sentence to the manual, that the the knob, faders and jog wheels calibration data are stored in the controller itself and therefore affect Mixxx 
- Copy the tables with the documentation of the HID FeatureReports into a comment section of Traktor-Kontrol-S2-MK1-hid-scripts.js

Than everything should be ready for merging this into the 2.3 branch. As soon as 2.3 is merged into Main, you could start with a new PR, that adds the calibration.

**Comment by leifhelm** [comment]
Have I understood correctly: The calibration feature will be only possible in the 2.4.0 release, but I should include a sentence about it in the manual???

**Comment by JoergAtGithub** [comment]
Calibration within Mixxx would require writing HID FeatureReports. And the support for HID FeatureReports is new in 2.4.

But calibration with the National Instruments Controler Editor is possible now, and will give users of Mixxx 2.3 calibrated controls. Therefore it should be mentioned in the manual of this mapping in the 2.3 version.

**Comment by leifhelm** [comment]
The calibration data would need to be read by mixxx and mixxx has to apply the correction. The values reported by the controller are raw measurements w/o calibration.

**Comment by JoergAtGithub** [comment]
> The calibration data would need to be read by mixxx and mixxx has to apply the correction. The values reported by the controller are raw measurements w/o calibration.

In this case, it should be documented of cause.

**Comment by leifhelm** [comment]
I explained it now.

**Comment by JoergAtGithub** [comment]
CI fails are unrelated:

- Pre-Commit fails due to missing Gitlab credentials #529 
- Sphinx / Link Check reports only links in untouched parts of the manual

**Pull request merged**